File: src/clients/MedEasy.Web/Scripts/MedEasy/RestObjects/FormField.ts
import { FormFieldType } from "../RestObjects/FormFieldType";
/**
 * Defines a form field and its content
 */
export class FormField {

    private _enabled: boolean | null;
    private _name: string | null;
    private _placeholder: string | null;
    private _type: FormFieldType;
    private _required: boolean | null;
    private _secret: boolean | null;
    private _minLength: number | null;
    private _maxLength: number | null;
    private _label: string | null;
    private _pattern: string | null;


    /** 
     * indicates whether or not the field value may be modified or submitted to a linked resource location. 
     * 
     **/
    public get enabled(): boolean | null {
        return this._enabled;
    }

    /** 
     * Type of the field 
    **/
    public get type(): FormFieldType | null {
        return this._type;
    }

    /** 
     * Label that can be associated with the field
     **/
    public get label(): string | null {
        return this._label;
    }

    /** 
     *  Name of the field that should be submitted
    **/
    public get name(): string | null {
        return this._name;
    }

    /** 
     *  Regular expression that the field should be validated against.
    **/
    public get pattern(): string | null {
        return this._pattern;
    }

    /** 
     *  Short hint that described the expected value of the field.
    **/
    public get placeholder(): string | null {
        return this._placeholder;
    }

    /** 
     *  Indicates if the field must be submitted
    **/
    public get required(): boolean | null {
        return this._required;
    }

    /** 
     *  Indicates the maximum length of the value
    **/
    public get maxLength(): number | null {
        return this._maxLength;
    }

    /** 
     *  Indicates the minimum length of the input
     **/
    public get minLength(): number | null {
        return this._minLength;
    }


    /** 
     *  Indicates whether or not the field value is considered sensitive information 
     *  and should be kept secret.
    **/
    public get secret(): boolean | null {
        return this._secret;
    }

    /** 
     *  Defines whether or not the field value is considered sensitive information 
     *  and should be kept secret.
     * @param {boolean | null} secret the new value
     **/
    public set secret(secret: boolean | null) {
        this._secret = secret;
    }


    /**
     * Buids a JQuery
     * @param {FormField} this. field of the form to build bootstrap element for.
     *
     * @returns the JQuery representation of the current element that can be append to a <form>
     */
    public create(): JQuery {
        let divField = $("<div class='form-group'>");

        let input: JQuery = $(`<input name='${this.name}' id='${this.name}'>`);
        switch (this.type) {
            case FormFieldType.Boolean:
                input.attr("type", "checkbox");
                break;
            case FormFieldType.Date:
                input.attr("type", "date")
                break;
            case FormFieldType.DateTime:
                input.attr("type", "datetime-local")
                break
            case FormFieldType.File:
                input.attr("type", "file");
                break;
            default:
                input.attr("type", "text");
                break;

        }

        if (!Boolean(this.enabled)) {
            input.prop("disabled", "disabled");
        }

        if (this.minLength) {
            input.attr("minlength", this.minLength);
        }

        if (this.maxLength) {
            input.attr("maxlength", this.maxLength);
        }


        let label: JQuery | null = null;
        if (this.label) {
            label = $(`<label for='${this.name}'>`);
        }
        if (this.type === FormFieldType.Boolean) {
            divField.append(input);
            if (label) {
                divField.append(label);
            }
        } else {
            if (label) {
                divField.append(label);
            }
            divField.append(input)
        }

        return divField;
    };
}

File: src/tools/MedEasy.RestObjects/FormField.cs
#if NETSTANDARD1_1
using Newtonsoft.Json;
using Newtonsoft.Json.Converters;
#else
using System.Text.Json.Serialization;
#endif
namespace MedEasy.RestObjects
{
    /// <summary>
    /// Form field representation
    /// </summary>
    /// <remarks>
    ///     Inspired by ION spec (see http://ionwg.org/draft-ion.html#form-fields for more details)
    /// </remarks>
#if NETSTANDARD1_1
    [JsonObject]
#endif
    public class FormField
    {
        /// <summary>
        /// indicates whether or not the field value may be modified or submitted to a linked resource location. 
        /// </summary>
        /// <remarks>
        /// 
        /// </remarks>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public bool? Enabled { get; set; }

        /// <summary>
        /// Type of the field 
        /// </summary>
#if NETSTANDARD1_1
        [JsonConverter(typeof(StringEnumConverter))]
#else
        [JsonConverter(typeof(JsonStringEnumConverter))]
#endif
        public FormFieldType Type { get; set; }

        /// <summary>
        /// Description of the field
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public string Label { get; set; }

        /// <summary>
        /// Name of the field that should be submitted
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public string Name { get; set; }

        /// <summary>
        /// Regular expression that the field should be validated against.
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public string Pattern { get; set; }

        /// <summary>
        /// Short hint that described the expected value of the field.
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public string Placeholder { get; set; }

        /// <summary>
        /// Indicates if the field must be submitted
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public bool? Required { get; set; }

        /// <summary>
        /// Indicates the maximum length of the value
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public int? MaxLength { get; set; }

        /// <summary>
        /// Indicates the minimum length of the value
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public int? MinLength { get; set; }

        /// <summary>
        /// Indicates that <see cref="Value"/>value must be greater than or equal to the specified <see cref="Min"/> value
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif        
        public int? Min { get; set; }

        /// <summary>
        /// Indicates that <see cref="Value"/>value must be less than or equal to the specified <see cref="Max"/> value
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public int? Max { get; set; }

        /// <summary>
        /// Indicates whether or not the field value is considered sensitive information 
        /// and should be kept secret.
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public bool? Secret { get; set; }

        /// <summary>
        /// a string description of the field that may be used to enhance usability.
        /// </summary>
#if NETSTANDARD1_1
        [JsonProperty(NullValueHandling = NullValueHandling.Ignore)]
#endif
        public string Description { get; set; }
    }
}